src/components/tabset/tab/tab.component.ts

```diff
@@ -10,13 +11,18 @@ let uniqueTabId = 0;
     templateUrl: './tab.component.html',
     changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class TabComponent implements OnDestroy {
+export class TabComponent implements OnInit, OnDestroy {
 
     /** Define the tab unique id */
     @Input() id: string = `ux-tab-${++uniqueTabId}`;
 
     /** Define the active state of this tab */
-    @Input() active: boolean = false;
+    @Input()
+    set active(active: boolean) {
+        if (active) {
+            this._tabset.setTabActive(this);
```

Comment:
Just noticed one more case where I still get an expression changed error. See the plunker, on initial load press the button to select the second tab:

https://plnkr.co/edit/Wh4yzMnwocPKNT8q?preview

[![Image from Gyazo](https://i.gyazo.com/d634c30ea83d75e7c4cf8dff9a1edaee.gif)](https://gyazo.com/d634c30ea83d75e7c4cf8dff9a1edaee)

Reply:
Do you think using BehaviorSubject is the wrong approach here? We seem to have this problem every time we use rxjs in components. I've added in `tick` and it solves the problem as usual, but it feels like a hack to trigger an unnecessary extra round of change detection like that.

Reply:
Perhaps moving it to `ngOnChanges` may help in this case? I'm not 100% sure this will fix the issue, but it could be worth a try. I agree that we seem to be having this problem quite a lot with setters, and perhaps it may be better to avoid them where possible.

Reply:
Unfortunately it's the same problem. I've also experimented with removing the observable altogether but there are still expression changed error.

Everything seems to work with the `tick` solution except for two karma tests:
```
HeadlessChrome 81.0.4044 (Windows 10.0.0) Tabset Component should change to new tab when tab is programmatically selected by tab index FAILED
        Error: Expected 'Schedule' to equal 'Solution'.
```
and
```
HeadlessChrome 81.0.4044 (Windows 10.0.0) Tabset Component should change to new tab when tab is programmatically selected by tab instance FAILED
        Error: Expected 'Schedule' to equal 'Solution'.
```

The usual `detectChanges`/`whenStable` routine does not seem to work in this case, which I don't really understand. The only way I got that assertion to pass is to put it behind a 10ms `setTimeout`. Do you have any ideas?

Reply:
I'm guessing it is because the tick operator makes the change asynchronous, so it will be evaluated on the next Javascript cycle, whereas the test is expecting it to be synchronous. You should just be able to wrap the test in `fakeAsync` and call the `tick` function, eg:

```typescript
it('should emit activeChange output when tab is selected', fakeAsync(async () => {
    await clickTab(2);

    tick(1); // we may not even need the 1

   // expectation here
}));
```

Reply:
For some reason it needs an actual time delay, 100ms worked where 1ms did not.
```ts
    fit('should change to new tab when tab is programmatically selected by tab index', fakeAsync(async () => {

        const tab1 = getTabItem(0);
        const tab3 = getTabItem(2);

        // check first tab is active
        expect(tab1.classList.contains('active')).toBeTruthy();

        spyOn(component, 'onTabActivated');

        // programatically change tab by index
        component.tabset.selectTab(2);
        tick(100);

        // check second tab is now active/first tab is not active
        expect(tab3.classList.contains('active')).toBeTruthy();
        expect(tab1.classList.contains('active')).toBeFalsy();

        // check tab contains correct content
        expect(nativeElement.querySelector('.tab-pane[aria-hidden="false"] h4').innerHTML).toEqual('Solution');

        expect(component.onTabActivated).toHaveBeenCalled();
    }));
```

If it's just the next cycle surely it should not need to wait a specific amount of time?

Reply:
Hmm, actually I thought the tick operator was using `Promise.resolve` which would fire on the next cycle, however it turns out it is using `requestAnimationFrame` so that could be a variable amount of time, but at 60fps it would likely be ~17ms, so that might be why there needs to be a delay?

Reply:
Thank you, it does seem reliable if not ideal for a unit-type test.